table/table.go

```diff
@@ -0,0 +1,532 @@
+package table
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/charmbracelet/lipgloss"
+	"github.com/mattn/go-runewidth"
+)
+
+// StyleFunc is the style function that determines the style of a Cell.
+//
+// It takes the row and column of the cell as an input and determines the
+// lipgloss Style to use for that cell position.
+//
+// Example:
+//
+//	t := table.New().
+//	    Headers("Name", "Age").
+//	    Row("Kini", 4).
+//	    Row("Eli", 1).
+//	    Row("Iris", 102).
+//	    StyleFunc(func(row, col int) lipgloss.Style {
+//	        switch {
+//	           case row == 0:
+//	               return HeaderStyle
+//	           case row%2 == 0:
+//	               return EvenRowStyle
+//	           default:
+//	               return OddRowStyle
+//	           }
+//	    })
+type StyleFunc func(row, col int) lipgloss.Style
+
+// DefaultStyles is a TableStyleFunc that returns a new Style with no attributes.
+func DefaultStyles(_, _ int) lipgloss.Style {
+	return lipgloss.NewStyle()
+}
+
+// Table is a type for rendering tables.
+type Table struct {
+	styleFunc StyleFunc
+	border    lipgloss.Border
+
+	borderTop    bool
+	borderBottom bool
+	borderLeft   bool
+	borderRight  bool
+	borderHeader bool
+	borderColumn bool
+	borderRow    bool
+
+	borderStyle lipgloss.Style
+	headers     []any
+	data        Data
+
+	width  int
+	height int
+	offset int
+
+	// widths tracks the width of each column.
+	widths []int
+
+	// heights tracks the height of each row.
+	heights []int
+}
+
+// New returns a new Table that can be modified through different
+// attributes.
+//
+// By default, a table has no border, no styling, and no rows.
+func New() *Table {
+	return &Table{
+		styleFunc:    DefaultStyles,
+		border:       lipgloss.RoundedBorder(),
+		borderBottom: true,
+		borderColumn: true,
+		borderHeader: true,
+		borderLeft:   true,
+		borderRight:  true,
+		borderTop:    true,
+	}
+}
+
+// ClearRows clears the table rows.
+func (t *Table) ClearRows() *Table {
+	t.data = nil
+	return t
+}
+
+// StyleFunc sets the style for a cell based on it's position (row, column).
+func (t *Table) StyleFunc(style StyleFunc) *Table {
+	t.styleFunc = style
+	return t
+}
+
+// style returns the style for a cell based on it's position (row, column).
+func (t *Table) style(row, col int) lipgloss.Style {
+	if t.styleFunc == nil {
+		return lipgloss.NewStyle()
+	}
+	return t.styleFunc(row, col)
+}
+
+// Data sets the table data.
+func (t *Table) Data(data Data) *Table {
+	t.data = data
+	return t
+}
+
+// Rows appends rows to the table data.
+func (t *Table) Rows(rows ...[]string) *Table {
+	if t.data == nil {
+		t.data = NewStringData()
+	}
+	for _, row := range rows {
+		switch t.data.(type) {
+		case *StringData:
+			t.data.(*StringData).Append(StringRow(row))
+		}
```

Comment:
@muesli does this seem reasonable to you? This is how we would avoid `Append` being a required message, but not sure if this is the correct way to do it.